### PR TITLE
Drop java 11, bump GraalVM to 23.0

### PR DESCRIPTION
**Pull request by bioball**
Note: I enabled build native jobs here because this PRB introduces changes there. I'm planning on getting rid of them again before merging.

A remaining TODO here is to tune our specialization limits (see commit for details)

After dropping Java 11, we can start using records, and also switch our Java code generators to use records.

### CODE REVIEW

File: buildSrc/src/main/kotlin/BuildInfo.kt [hunk 1]
```diff
@@ -20,47 +18,25 @@ open class BuildInfo(project: Project) {
...
-    }
-
-    val arch by lazy {
-      if (os.isMacOsX && isGraal22) {
-        "amd64"
-      } else {
-        self.arch
-      }
+    val graalVmJavaVersion: String by lazy {
+      libs.findVersion("graalVMJdkVersion").get().toString()
```
Comment (translatenix) [inline buildSrc/src/main/kotlin/BuildInfo.kt:22]: nit: `graalVmJavaVersion` vs. `graalVMJdkVersion` (two differences)

File: pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeMethodVirtualNode.java [hunk 2]
```diff
@@ -102,7 +103,7 @@ protected Object evalFunction(
       VirtualFrame frame,
       VmFunction receiver,
       VmClass receiverClass,
-      @Cached("create()") IndirectCallNode callNode) {
+      @Exclusive @Cached(value = "create()") IndirectCallNode callNode) {
```
Comment (translatenix) [inline pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeMethodVirtualNode.java:106]: `value =` is redundant

File: pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeMethodVirtualNode.java [hunk 3]
```diff
@@ -142,7 +143,7 @@ protected Object eval(
       VirtualFrame frame,
       Object receiver,
       VmClass receiverClass,
-      @Cached("create()") IndirectCallNode callNode) {
+      @Exclusive @Cached(value = "create()") IndirectCallNode callNode) {
```
Comment (translatenix) [inline pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeMethodVirtualNode.java:146]: `value =` is redundant

File: pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeSuperMethodNode.java [hunk 4]
```diff
@@ -51,8 +51,8 @@ protected InvokeSuperMethodNode(
   @Specialization
   protected Object eval(
       VirtualFrame frame,
-      @Cached("findSupermethod(frame)") ClassMethod supermethod,
-      @Cached("create(supermethod.getCallTarget(sourceSection))") DirectCallNode callNode) {
+      @Cached(value = "findSupermethod(frame)", neverDefault = true) ClassMethod supermethod,
+      @Cached(value = "create(supermethod.getCallTarget(sourceSection))") DirectCallNode callNode) {
```
Comment (translatenix) [inline pkl-core/src/main/java/org/pkl/core/ast/expression/member/InvokeSuperMethodNode.java:55]: `value =` is redundant

**Review by translatenix — COMMENTED** [review]
noticed a few nits

File: buildSrc/src/main/kotlin/BuildInfo.kt (hunk 1 above)
Comment (bioball) [reply buildSrc/src/main/kotlin/BuildInfo.kt:22]: Yeah, fair. Going to split the difference here (`graalVmJdkVersion`)

**Review by holzensp — APPROVED** [review]

File: bench/bench.gradle.kts [hunk 5]
```diff
@@ -38,7 +38,7 @@ jmh {
 }
 
 tasks.named("jmh") {
-  dependsOn(":installGraalVm")
+  dependsOn(":installGraalVmAmd64")
```
[the remark this answers is not in the document]
Comment (bioball) [reply bench/bench.gradle.kts:41]: I think I'll leave this as a TODO. Would certainly be nice to have benchmarks for both architectures.

**Pull request merged**